File: 프로그래머스/2/49994. 방문 길이/방문 길이.cs
using System;

public class Solution {
    private int[] dx = {1, -1, 0, 0};
    private int[] dy = {0, 0, 1, -1};
    public int solution(string dirs) {
        bool[,,] visited = new bool[11, 11, 4];
        
        int x = 5; int y = 5; int d = 0;
        int answer = 0;
        
        foreach(char c in dirs.ToCharArray())
        {
            int opposite_d = 0;

            if(c == 'R') { d = 0; opposite_d = 1; }
            if(c == 'L') { d = 1; opposite_d = 0; }
            if(c == 'D') { d = 2; opposite_d = 3; }
            if(c == 'U') { d = 3; opposite_d = 2; }
            
            int nx = x + dx[d];
            int ny = y + dy[d];
            if (nx < 0 || nx >= 11 || ny < 0 || ny >= 11) continue;

            if(!visited[x, y, d])
            {
                visited[x, y, d] = true;
                visited[nx, ny, opposite_d ] = true;
                answer++;
            }
            x = nx;
            y = ny;
        }
        return answer;
    }
}

File: 프로그래머스/2/49994. 방문 길이/방문 길이.cpp
#include <string>
#include <bits/stdc++.h>
using namespace std;


int dy[] = {-1, 0, 1, 0};
int dx[] = {0, 1, 0, -1};

int solution(string dirs) {
    int answer = 0;
    int y = 0, x = 0;
    
    set<string> visited;
    
    for(int i = 0; i < dirs.size(); i++)
    {
        char c = dirs[i];
        int dir = 0;
        
        if(c == 'D') dir = 0;
        else if(c == 'R') dir = 1;
        else if(c == 'U') dir = 2;
        else dir = 3;
        
        int ny = y + dy[dir];
        int nx = x + dx[dir];
        
        if(ny < -5 || ny > 5 || nx < -5 || nx > 5) continue;
        
        string path1 = to_string(y) + "," + to_string(x) + "," + to_string(ny) + "," + to_string(nx);
        string path2 = to_string(ny) + "," + to_string(nx) + "," + to_string(y) + "," + to_string(x);
        
        if(visited.find(path1) == visited.end())
        {
            visited.insert(path1);
            visited.insert(path2);
            answer++;
        }
        y = ny;
        x = nx;
    }
    return answer;
}